folium/map.py

```diff
@@ -383,7 +400,11 @@ def __init__(
         self.options = remove_empty(
             draggable=draggable or None, autoPan=draggable or None, **kwargs
         )
+        self.icon: Optional[Icon] = None
         if icon is not None:
+            # here we make sure the icon is rendered after the last marker that needs it
+            if icon._parent is not None:
+                del icon._parent._children[icon.get_name()]
             self.add_child(icon)
```

Comment:
In my proposal, the Icon could be added to the map (or to the parent of the Marker). We can keep this line to avoid breaking backwards compatibility. I am not sure if `add_child` is idempotent. In general I think for `Folium` at least it should be.